From 12ecaea1ffffb956e61491e5f8cedbd8cf01d093 Mon Sep 17 00:00:00 2001
From: Baji Patthan <bpatth@codeaurora.org>
Date: Tue, 5 May 2015 19:38:46 -0700
Subject: [PATCH 14/22] msm: usb: Add support for SBC8016

Add support for dynamic switch between device mode and host mode on the
same USB port.
The following code implements switch between the HOST
mode to device mode when used different HW components
on the same port: USB HUB and the USB jack type B for device mode.

Change-Id: Ib5460cce761d2af813054d2018f629a4cb292429
Signed-off-by: Baji Patthan <bpatth@codeaurora.org>
---
 drivers/usb/phy/phy-msm-usb.c | 22 +++++++++++++++++++++-
 include/linux/usb/msm_hsusb.h |  2 ++
 2 files changed, 23 insertions(+), 1 deletion(-)

diff --git a/drivers/usb/phy/phy-msm-usb.c b/drivers/usb/phy/phy-msm-usb.c
index 1ffa6d6..7a81430 100644
--- a/drivers/usb/phy/phy-msm-usb.c
+++ b/drivers/usb/phy/phy-msm-usb.c
@@ -3805,12 +3805,16 @@ static void msm_id_status_w(struct work_struct *w)
 	else if (motg->ext_id_irq)
 		id_state = gpio_get_value(motg->pdata->usb_id_gpio);
 
-	if (id_state) {
+	if (motg->id_state) {
+		if (motg->pdata->switch_sel_gpio > 0)
+			gpio_direction_input(motg->pdata->switch_sel_gpio);
 		if (!test_and_set_bit(ID, &motg->inputs)) {
 			pr_debug("ID set\n");
 			work = 1;
 		}
 	} else {
+		if (motg->pdata->switch_sel_gpio > 0)
+			gpio_direction_output(motg->pdata->switch_sel_gpio, 1);
 		if (test_and_clear_bit(ID, &motg->inputs)) {
 			pr_debug("ID clear\n");
 			set_bit(A_BUS_REQ, &motg->inputs);
@@ -4795,9 +4799,25 @@ struct msm_otg_platform_data *msm_otg_dt_to_pdata(struct platform_device *pdev)
 	if (pdata->pmic_id_irq < 0)
 		pdata->pmic_id_irq = 0;
 
+	pdata->hub_reset_gpio = of_get_named_gpio(node, "qcom,hub-reset-gpio", 0);
+	if (pdata->hub_reset_gpio < 0)
+		pr_debug("hub_reset_gpio is not available\n");
+	else
+		gpio_direction_output(pdata->hub_reset_gpio, 1);
+
+	pdata->switch_sel_gpio = of_get_named_gpio(node, "qcom,sw-sel-gpio", 0);
+	if (pdata->switch_sel_gpio < 0)
+		pr_debug("switch_sel_gpio is not available\n");
+
 	pdata->usb_id_gpio = of_get_named_gpio(node, "qcom,usbid-gpio", 0);
 	if (pdata->usb_id_gpio < 0)
 		pr_debug("usb_id_gpio is not available\n");
+	else {
+		if (gpio_get_value(pdata->usb_id_gpio))
+			gpio_direction_input(pdata->switch_sel_gpio);
+		else
+			gpio_direction_output(pdata->switch_sel_gpio, 1);
+	}
 
 	pdata->l1_supported = of_property_read_bool(node,
 				"qcom,hsusb-l1-supported");
diff --git a/include/linux/usb/msm_hsusb.h b/include/linux/usb/msm_hsusb.h
index d3a3401..e5d52c2 100644
--- a/include/linux/usb/msm_hsusb.h
+++ b/include/linux/usb/msm_hsusb.h
@@ -298,6 +298,8 @@ struct msm_otg_platform_data {
 	bool enable_ahb2ahb_bypass;
 	bool disable_retention_with_vdd_min;
 	int usb_id_gpio;
+	int hub_reset_gpio;
+	int switch_sel_gpio;
 	bool phy_dvdd_always_on;
 };
 
-- 
1.8.2.1

